pulsar-transaction/buffer/src/main/java/org/apache/pulsar/transaction/buffer/impl/TransactionCursorImpl.java

```diff
@@ -0,0 +1,142 @@
+import org.apache.pulsar.transaction.buffer.TransactionCursor;
+import org.apache.pulsar.transaction.buffer.TransactionMeta;
+import org.apache.pulsar.transaction.buffer.exceptions.NoTxnsCommittedAtLedgerException;
+import org.apache.pulsar.transaction.buffer.exceptions.TransactionNotFoundException;
+import org.apache.pulsar.transaction.impl.common.TxnID;
+
+public class TransactionCursorImpl implements TransactionCursor {
+
+
+    private ConcurrentMap<TxnID, TransactionMetaImpl> txnIndex;
+    private Map<Long, Set<TxnID>> committedTxnIndex;
+
+    TransactionCursorImpl() {
+        this.txnIndex = new ConcurrentHashMap<>();
+        this.committedTxnIndex = new TreeMap<>();
+    }
+
+    @Override
+    public CompletableFuture<TransactionMeta> getTxnMeta(TxnID txnID, boolean createIfNotExist) {
+        CompletableFuture<TransactionMeta> getFuture = new CompletableFuture<>();
+        TransactionMeta meta = txnIndex.get(txnID);
+        if (null == meta) {
+            if (!createIfNotExist) {
+                getFuture.completeExceptionally(
+                    new TransactionNotFoundException("Transaction `" + txnID + "` doesn't" + " exist"));
+                return getFuture;
+            }
+
+            TransactionMetaImpl newMeta = new TransactionMetaImpl(txnID);
+            txnIndex.putIfAbsent(txnID, newMeta);
+            meta = newMeta;
+        }
+        getFuture.complete(meta);
+
+        return getFuture;
+    }
+
+    @Override
+    public CompletableFuture<Void> commitTxn(long committedLedgerId, long committedEntryId, TxnID txnID,
+                                             Position position) {
+        CompletableFuture<Void> commitFuture = new CompletableFuture<>();
+        TransactionMeta meta = getMeta(txnID, commitFuture);
+        if (commitFuture.isCompletedExceptionally()) {
+            return commitFuture;
+        }
+
+        synchronized (meta) {
+            commitFuture = meta.commitTxn(committedLedgerId, committedEntryId)
+                               .thenAccept(ignore -> addTxnToCommittedIndex(txnID, committedLedgerId));
+        }
+
+        return commitFuture;
+    }
+
+    private void addTxnToCommittedIndex(TxnID txnID, long committedAtLedgerId) {
+        synchronized (committedTxnIndex) {
+            committedTxnIndex.computeIfAbsent(committedAtLedgerId, ledgerId -> new HashSet<>()).add(txnID);
+        }
+    }
+
+    @Override
+    public CompletableFuture<Void> abortTxn(TxnID txnID) {
+        CompletableFuture<Void> abortFuture = new CompletableFuture<>();
+
+        TransactionMeta meta = getMeta(txnID, abortFuture);
+        if (abortFuture.isCompletedExceptionally()) {
+            return abortFuture;
+        }
+
+        synchronized (meta) {
+            abortFuture = meta.abortTxn().thenApply(ignore -> null);
+        }
+
+        return abortFuture;
+    }
+
+    private TransactionMeta getMeta(TxnID txnID, CompletableFuture<Void> future) {
```

Comment:
This method is so awkward.

Change it to `CompletableFuture<TransactionMeta> getMeta(TxnID txnID)`